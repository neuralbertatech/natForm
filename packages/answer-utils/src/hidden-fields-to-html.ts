import { HiddenFieldAnswer } from '@heyform-inc/shared-types-enums'

export function hiddenFieldsToHtml(hiddenFields: HiddenFieldAnswer[]): string {
  if (!hiddenFields.length) return ''

  const html = hiddenFields
    .map(hiddenField => {
      return `
<li>
  <h3>${hiddenField.name}</h3>
  <p>${hiddenField.value}</p>
</li>
`
    })
    .join('')

  return `<ol>${html}</ol>`
}
