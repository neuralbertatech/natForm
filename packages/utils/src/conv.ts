import { IParseOptions, IStringifyOptions, parse, stringify } from 'qs'
import { convert } from 'html-to-text'

import {
  isEmpty,
  isBoolean,
  isTrue,
  isNumber,
  isValid,
  isString
} from './helper'

export function toBool(value: unknown, defaults?: boolean): boolean {
  if (isEmpty(value)) {
    return defaults || false
  }

  if (isBoolean(value)) {
    return value as boolean
  }

  return isTrue(value)
}

export const parseBool = toBool

export function toInteger(
  value: unknown,
  defaults?: number,
  maxValue?: number
): number | undefined {
  let val: number

  if (isNumber(value)) {
    val = Number(value)
  } else {
    val = parseInt(value as string, 10)
  }

  if (!isFinite(val)) {
    return defaults
  }

  return maxValue ? Math.min(maxValue, val) : val
}

export const parseNumber = toInteger
export const toInt = toInteger

export function toFloat(
  value: unknown,
  defaults?: number,
  maxValue?: number
): number | undefined {
  let val: number

  if (isNumber(value)) {
    val = Number(value)
  } else {
    val = parseFloat(value as string)
  }

  if (!isFinite(val)) {
    return defaults
  }

  return maxValue ? Math.min(maxValue, val) : val
}

export function toFixed(value: number, precision = 2) {
  if (isNumber(value)) {
    return value.toFixed(precision).replace(/(\.\d*?)0+$/, '$1').replace(/\.$/, '')
  }
}

export function toJSON<T extends object>(
  str: unknown,
  defaults?: T
): T | undefined {
  let value: T | undefined

  if (isValid(str) && isString(str)) {
    try {
      value = JSON.parse(str as string)
    } catch (e) {
      // eslint-disable-line
    }
  }

  if (!!defaults && !value) {
    value = defaults
  }

  return value
}

export const parseJson = toJSON

const THOUSAND = 1_000
const HUNDRED_THOUSAND = 100_000
const MILLION = 1_000_000
const HUNDRED_MILLION = 100_000_000
const BILLION = 1_000_000_000
const HUNDRED_BILLION = 100_000_000_000
const TRILLION = 1_000_000_000_000

export function toIntlNumber(value: number) {
  if (value >= THOUSAND && value < MILLION) {
    const thousands = value / THOUSAND

    if (thousands === Math.floor(thousands) || value >= HUNDRED_THOUSAND) {
      return Math.floor(thousands) + 'K'
    } else {
      return Math.floor(thousands * 10) / 10 + 'K'
    }
  } else if (value >= MILLION && value < BILLION) {
    const millions = value / MILLION

    if (millions === Math.floor(millions) || value >= HUNDRED_MILLION) {
      return Math.floor(millions) + 'M'
    } else {
      return Math.floor(millions * 10) / 10 + 'M'
    }
  } else if (value >= BILLION && value < TRILLION) {
    const billions = value / BILLION

    if (billions === Math.floor(billions) || value >= HUNDRED_BILLION) {
      return Math.floor(billions) + 'B'
    } else {
      return Math.floor(billions * 10) / 10 + 'B'
    }
  } else {
    return `${value}`
  }
}

interface ToDurationOptions {
  hideOnZeroValue?: boolean
  padNumber?: boolean
  hourUnit?: string
  minuteUnit?: string
  secondUnit?: string
}

function durationPad(num: number, size: number) {
  return ('000' + num).slice(size * -1)
}

function getDuration(num: number, unit: string, options?: ToDurationOptions) {
  if (options?.hideOnZeroValue && num === 0) {
    return
  }

  if (options?.padNumber) {
    return durationPad(num, 2) + unit
  }

  return num + unit
}

export function toDuration(value: number, options?: ToDurationOptions) {
  const hours = Math.floor(value / 60 / 60)
  const minutes = Math.floor(value / 60) % 60
  const seconds = Math.floor(value - minutes * 60 - hours * 60 * 60)

  return [
    hours > 0 ? { num: hours, unit: options?.hourUnit || 'h' } : null,
    minutes > 0 ? { num: minutes, unit: options?.minuteUnit || 'm' } : null,
    { num: seconds, unit: options?.secondUnit || 's' }
  ]
    .filter(Boolean)
    .map((d, i) =>
      getDuration(d!.num, d!.unit, {
        ...options,
        hideOnZeroValue: i === 0 ? false : options?.hideOnZeroValue
      })
    )
    .filter(Boolean)
    .join(' ')
}

export function toURLParams<T extends object>(
  value: string,
  options?: IParseOptions
) {
  return parse(value, options) as T
}

export function toURLQuery(
  value: Record<string | number, any>,
  baseUri?: string,
  options?: IStringifyOptions
) {
  const query = stringify(value, options)

  if (baseUri) {
    return baseUri + (baseUri.includes('?') ? '&' : '?') + query
  }

  return query
}

export function htmlToText(html: string, limit = 100) {
  const value = convert(html, {
    selectors: [
      { selector: 'a', options: { ignoreHref: true } },
      { selector: 'img', format: 'skip' },
      { selector: 'picture', format: 'skip' },
      { selector: 'video', format: 'skip' },
      { selector: 'audio', format: 'skip' },
      { selector: 'iframe', format: 'skip' }
    ],
  }).replace(/\n/g, ' ')

  if (limit > 0 && value.length > limit) {
    return value.slice(0, limit)
  }

  return value
}
