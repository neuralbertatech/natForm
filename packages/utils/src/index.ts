export * from 'deep-object-diff'
export { default as qs } from 'qs'
export * from './bytes'
export * from './clone'
export * from './color'
export * from './conv'
export * from './date'
export { default as helper } from './helper'
export * from './second'
export * from './mime'
export * from './nanoid'
export * from './object'
export * from './random'
export * from './slugify'
export * from './type'
export * from './uuid'
