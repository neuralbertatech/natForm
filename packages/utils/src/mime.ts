import db from './mime-db.json'
import { isEmpty, isString } from './helper'

export function mime(path: string) {
  if (isEmpty(path) || !isString(path) || path.length > 100) {
    return
  }

  // Fork from https://github.com/broofa/mime/blob/a34dac529c2aaf2a0bc5ff8be26f983795d0a5e9/src/Mime.ts#L73
  // Remove chars preceeding `/` or `\`
  const last = path.replace(/^.*[/\\]/, '').toLowerCase()

  // Remove chars preceeding '.'
  const ext = last.replace(/^.*\./, '').toLowerCase()

  const hasPath = last.length < path.length
  const hasDot = ext.length < last.length - 1

  // Extension-less file?
  if (!hasDot && hasPath) {
    return
  }

  return (db as Record<string, string>)[ext]
}

export const commonImageMimeTypes = [
  mime('.jpg'),
  mime('.png'),
  mime('.bmp'),
  mime('.webp'),
  mime('.gif')
] as unknown as string[]

export const commonFileMimeTypes = [
  ...commonImageMimeTypes,
  mime('.txt'),
  mime('.md'),
  mime('.doc'),
  mime('.docx'),
  mime('.xls'),
  mime('.xlsx'),
  mime('.csv'),
  mime('.ppt'),
  mime('.pptx'),
  mime('.pdf'),
  mime('.mp4'),
  mime('.wmv'),
  mime('.svg')
] as unknown as string[]

export const IMAGE_MIME_TYPES = commonImageMimeTypes
export const FILE_MIME_TYPES = commonFileMimeTypes
