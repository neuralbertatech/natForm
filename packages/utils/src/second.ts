import _ms from 'ms'

export function hs(arg: string): number | undefined {
  const value = ms(arg)

  if (value) {
    return Math.round(value / 1_000)
  }
}

export function ms(arg: string): number | undefined {
  return _ms(arg)
}

export const toSecond = hs
export const toMillisecond = ms
