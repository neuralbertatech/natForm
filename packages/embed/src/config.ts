import { AnyMap, EmbedConfig } from './type'
import { $ } from './utils'

const ATTR_PREFIX = 'data-heyform-'
const HIDDEN_FIELD_PREFIX = `${ATTR_PREFIX}hiddenfield-`
const ID_ATTR_NAME = `${ATTR_PREFIX}id`
const TYPE_ATTR_NAME = `${ATTR_PREFIX}type`

export function getConfigs() {
  const $form = $(`[${ID_ATTR_NAME}]`)

  return $form.map(el => {
    const settings: AnyMap = {}
    const hiddenFields: AnyMap = {}

    const names = el.getAttributeNames()

    names.forEach(name => {
      let key = name.toLowerCase()

      if (key.startsWith(HIDDEN_FIELD_PREFIX)) {
        key = key.replace(HIDDEN_FIELD_PREFIX, '')

        hiddenFields[key] = el.getAttribute(name)
      } else if (key.startsWith(ATTR_PREFIX)) {
        key = key.replace(ATTR_PREFIX, '').replace(/(\-)+([a-z])/gi, (_, __, s) => s.toUpperCase())

        settings[key] = el.getAttribute(name)
      }
    })

    const formId = el.getAttribute(ID_ATTR_NAME) as string
    const type = el.getAttribute(TYPE_ATTR_NAME) as string

    return {
      formId,
      type,
      container: $(el as HTMLElement),
      settings,
      hiddenFields
    } as EmbedConfig<any>
  })
}
