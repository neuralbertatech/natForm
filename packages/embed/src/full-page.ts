import { Dom } from '@/utils'

import IconLoading from './assets/icon-loading.svg'
import { Standard } from './standard'
import { FullPageSettings } from './type'

const FULLPAGE_TEMPLATE = `
<div class="heyform__iframe-container">
  <iframe src="{src}" allow="microphone; camera"></iframe>
  <div class="heyform__loading-container">${IconLoading}</div>
</div>
`

export class FullPage extends Standard<FullPageSettings> {
  override render() {
    this.$container
      .style({
        position: 'fixed',
        top: 0,
        left: 0,
        width: '100%',
        height: '100%',
        zIndex: '9999'
      })
      .append(
        Dom.compile(FULLPAGE_TEMPLATE, {
          src: this.formUrl
        })
      )
  }
}
