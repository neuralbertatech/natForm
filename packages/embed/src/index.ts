import { getConfigs } from './config'
import { FullPage } from './full-page'
import { Modal } from './modal'
import { Popup } from './popup'
import { Standard } from './standard'
import './style.scss'
import { logger } from './utils'

const loaded = new Set<string>()
const instances = new Map<string, Modal<any> | Popup<any>>()

function main() {
  try {
    const configs = getConfigs()

    configs.forEach(c => {
      const key = `${c.formId}-${c.type}`

      if (!loaded.has(key)) {
        loaded.add(key)

        switch (c.type) {
          case 'modal':
            return instances.set(key, new Modal(c))
    
          case 'popup':
            return instances.set(key, new Popup(c))
    
          case 'fullpage':
            return new FullPage(c)
    
          default:
            return new Standard(c)
        }
      }
    })
  
    logger.info('Configs', JSON.stringify(configs))
  } catch (err) {
    logger.error(err)
  }
}

function open(formId: string, type: 'modal' | 'popup') {
  const instance = instances.get(`${formId}-${type}`)

  if (instance) {
    instance.open()
  }
}

function close(formId: string, type: 'modal' | 'popup') {
  const instance = instances.get(`${formId}-${type}`)

  if (instance) {
    instance.close()
  }
}

function toggle(formId: string, type: 'modal' | 'popup') {
  const instance = instances.get(`${formId}-${type}`)

  if (instance) {
    instance.toggle()
  }
}

export const openModal = (formId: string) => open(formId, 'modal')
export const closeModal = (formId: string) => close(formId, 'modal')
export const toggleModal = (formId: string) => toggle(formId, 'modal')
export const openPopup = (formId: string) => open(formId, 'popup')
export const closePopup = (formId: string) => close(formId, 'popup')
export const togglePopup = (formId: string) => toggle(formId, 'popup')

// Import the library after the element
main()

// Import the library before the element
window.addEventListener('DOMContentLoaded', main)
