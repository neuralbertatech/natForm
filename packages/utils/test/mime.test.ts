import { test, expect } from 'vitest'
import { mime, commonImageMimeTypes, commonFileMimeTypes } from '../src'

test('image mime type', () => {
  expect(mime('.jpg')).toBe('image/jpeg')
})

test('common image mime type', () => {
  expect(commonImageMimeTypes).toMatchSnapshot()
})

test('common file mime type', () => {
  expect(commonFileMimeTypes).toMatchSnapshot()
})

test('mime', () => {
  const configs = [
    // Bare extension
    { input: 'txt', expected: 'text/plain' },
    { input: '.txt', expected: 'text/plain' },
    { input: '.bogus', expected: undefined },
    { input: 'bogus', expected: undefined },

    // File paths
    { input: 'dir/text.txt', expected: 'text/plain' },
    { input: 'dir\\text.txt', expected: 'text/plain' },
    { input: '.text.txt', expected: 'text/plain' },
    { input: '.txt', expected: 'text/plain' },
    { input: 'txt', expected: 'text/plain' },
    { input: '/path/to/page.html', expected: 'text/html' },
    { input: 'c:\\path\\to\\page.html', expected: 'text/html' },
    { input: 'page.html', expected: 'text/html' },
    { input: 'path/to/page.html', expected: 'text/html' },
    { input: 'path\\to\\page.html', expected: 'text/html' },
    { input: '/txt', expected: undefined },
    { input: '\\txt', expected: undefined },
    { input: 'text.nope', expected: undefined },
    { input: '/path/to/file.bogus', expected: undefined },
    { input: '/path/to/json', expected: undefined },
    { input: '/path/to/.json', expected: undefined },
    { input: '/path/to/.config.json', expected: 'application/json' },
    { input: '.config.json', expected: 'application/json' },

    // Non-sensical
    { input: null, expected: undefined },
    { input: undefined, expected: undefined },
    { input: 42, expected: undefined },
    { input: {}, expected: undefined },
  ]

  configs.forEach(row => {
    expect(mime(row.input as any)).toBe(row.expected)
  })
})