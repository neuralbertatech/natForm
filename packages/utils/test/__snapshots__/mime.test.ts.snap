// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`common file mime type 1`] = `
[
  "image/jpeg",
  "image/png",
  "image/bmp",
  "image/webp",
  "image/gif",
  "text/plain",
  "text/markdown",
  "application/msword",
  "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
  "application/vnd.ms-excel",
  "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
  "text/csv",
  "application/vnd.ms-powerpoint",
  "application/vnd.openxmlformats-officedocument.presentationml.presentation",
  "application/pdf",
  "video/mp4",
  "video/x-ms-wmv",
  "image/svg+xml",
]
`;

exports[`common image mime type 1`] = `
[
  "image/jpeg",
  "image/png",
  "image/bmp",
  "image/webp",
  "image/gif",
]
`;
